pytorch_lightning/trainer/trainer_conf.py

```diff
@@ -0,0 +1,237 @@
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Optional
+
+from hydra.core.config_store import ConfigStore
+from hydra.types import ObjectConf
+from hydra.utils import instantiate
+from omegaconf import MISSING, DictConfig
+
+cs = ConfigStore.instance()
+
+
+@dataclass
+class LightningTrainerConf:
+    default_root_dir: Optional[str] = None
+    gradient_clip_val: float = 0
+    process_position: int = 0
+    num_nodes: int = 1
+    num_processes: int = 1
+    gpus: Optional[Any] = None
+    auto_select_gpus: bool = False
+    tpu_cores: Optional[Any] = None
+    log_gpu_memory: Optional[str] = None
+    progress_bar_refresh_rate: int = 1
+    overfit_batches: float = 0.0
+    track_grad_norm: Any = -1
+    check_val_every_n_epoch: int = 1
+    fast_dev_run: bool = False
+    accumulate_grad_batches: Any = 1
+    max_epochs: int = 1000
+    min_epochs: int = 1
+    max_steps: Optional[int] = None
+    min_steps: Optional[int] = None
+    limit_train_batches: float = 1.0
+    limit_val_batches: float = 1.0
+    limit_test_batches: float = 1.0
+    val_check_interval: float = 1.0
+    log_save_interval: int = 100
+    row_log_interval: int = 50
+    distributed_backend: Optional[str] = None
+    precision: int = 32
+    print_nan_grads: bool = False
+    weights_summary: Optional[str] = "top"
+    weights_save_path: Optional[str] = None
+    num_sanity_val_steps: int = 2
+    truncated_bptt_steps: Optional[int] = None
+    resume_from_checkpoint: Optional[str] = None
+    benchmark: bool = False
+    deterministic: bool = False
+    reload_dataloaders_every_epoch: bool = False
+    auto_lr_find: Any = False
+    replace_sampler_ddp: bool = True
+    terminate_on_nan: bool = False
+    auto_scale_batch_size: Any = False
+    prepare_data_per_node: bool = True
+    amp_level: str = "O1"
+    num_tpu_cores: Optional[int] = None
+
+
+cs.store(group="trainer", name="trainer", node=LightningTrainerConf)
```

Comment:
Please namespace all of the pl configs:
`pl/trainer`.

Reply:
@omry Can you please explain what are benefits of this? 

Won't this just only make the configuration deeper? and as a result when the user will like to override trainer parameters from command line will have to add `pl` everywhere, like `pl.trainer.gpus=[2,3]` etc.?

Asking as in NeMo (where we independently started to merge PT Lightning with Hydra like[ two weeks ago](https://github.com/NVIDIA/NeMo/pull/816);)) we just yesterday [did the opposite](https://github.com/NVIDIA/NeMo/pull/895) and got rid of `pl`...

Reply:
@tkornuta-nvidia, yes:
You can think of the configuration object Hydra is composing as a shared object that is potentially the home for more than one user.
As an example, the configuration is already hosting both Hydra itself under the hydra node (and the hydra/ config groups) and the user config.
Different frameworks that are installed can make their configs accessible via Hydra (via a Config Searchpath plugin and/or by storing configs in the ConfigStore like in the example above).
To avoid collisions between themselves and with the end user, I am encouraging framework owners to place their configs in some kind of namespace.

Imagine the following scenario:
A user wants to use NeMo and Fairseq at the same time assuming they both use Hydra.
If both conveniently places their model configs in the config group model and in the config node model, the user can now easily try to configure the NeMo model with a configuration for a Fairseq model.
namespacing is solving this.

This of it as the equivalent of Java packages or Python modules, but for configuration.
The cost that you are mentioning about the command line being longer is real, but I think It's still the right choice.